### PR TITLE
Allow successful empty response

### PR DESCRIPTION
**Pull request by j-frei**
Dear @vanderbilt-redcap team, @mmcev106,

Thank you for your FHIR-module for REDCap.
We'd like to evaluate and use your FHIR extension in order to process and store survey forms on our FHIR server.

As a sidenote, we currently use the IBM FHIR Server from here: 
https://github.com/LinuxForHealth/FHIR
formerly https://github.com/IBM/FHIR

When sending the FHIR Bundle to the remote FHIR server, we got error messages due to the fact that the module tries to parse the response from the FHIR server. In our case, the FHIR server acknowledges the POST request with the HTTP status code 201, but response with an empty body.
Since the module expects a valid JSON-encoded Bundle as a response body and tries to parse the body in order to verify the resource type of the FHIR response, an exception is thrown.

In this PR we include the case for empty response data in which the module regards the empty response as valid if the response HTTP code propagates a successful response in the success range (200 <= \<status code\> < 300).

I'd like to know your feedback on this PR and whether it is addressing the pain point adequately.

Kind regards,
Johann

### CODE REVIEW

**Comment by mmcev106** [comment]
@j-frei, thanks very much!  I plan to look at this PR in the next couple of weeks.

File: FHIRServicesExternalModule.php [hunk 1]
```diff
@@ -734,7 +734,9 @@ function toArray($fhirObjectOrArray){
             $handle($a);
             return $a;
         }
-        else{
+        else if(is_null($fhirObjectOrArray)){
+            return "null";
+        } else {
```
Comment (mmcev106) [inline FHIRServicesExternalModule.php]: `null` is not a valid `$fhirObjectOrArray` or Array.  Instead of this check, would it be more appropriate to check for `null` before the `toArray()` function is called?

File: FHIRServicesExternalModule.php [hunk 2]
```diff
@@ -2805,19 +2807,45 @@ function sendToRemoteFHIRServer($resource){
...
 
+        $successful_response_with_empty_body = false;
+        if(empty($response)){
+            // parse for response HTTP status code
+            $http_response_line = $http_response_header[0];
+            // look for "HTTP/1.1 [[STATUS_CODE]] OK"
+            if(preg_match("/\s([0-9]+)\s/", $http_response_line, $match)){
+                $http_response_status_code = intval($match[0]);
+                if($http_response_status_code>=200 && $http_response_status_code<300){
+                    $successful_response_with_empty_body = true;
```
Comment (mmcev106) [inline FHIRServicesExternalModule.php]: Would it be simpler but just as effective to return `null` here, and revert the changes below this line?  I don't think there's any reason to continue on to the parsing & type check if we already know the response is empty.

**Review by mmcev106 — CHANGES_REQUESTED** [review]
Functionally this makes sense.  I do have a couple of suggestions that I think would simplify things and make the module easier to maintain going forward.

File: FHIRServicesExternalModule.php (hunk 1 above)
Comment (j-frei) [reply FHIRServicesExternalModule.php]: I agree that the check can be moved to somewhere else.
The obvious option would be to move it the the caller `jsonSerialize`, since this function is the only place where `toArray` is called as far a I see.
Even though, the `jsonSerialize` function parameter is called `$fhirObjectOrArray` as well. Should we rename this parameter for maintenance/readability reasons to something like `$fhirObjectOrArrayOrNull`?
Moving the null-check even before the `jsonSerialize` is not useful in my opinion since the the function is called from nine different locations.
But, of course, I am interested in your opinion.

File: FHIRServicesExternalModule.php (hunk 2 above)
Comment (j-frei) [reply FHIRServicesExternalModule.php]: Yes, you are right.
My initial intention was to keep the dataflow equivalent for all cases and also let empty requests be processed through the `$this->parse($response);`-logic.
But we can also take a shortcut here and save multiple LoCs.

**Comment by j-frei** [comment]
Thank you for your feedback, Mark @mmcev106!
I have updated the the pull request to address your comments and improvement suggestions adequately.
Let me know whether the code changes make sense to you.

File: FHIRServicesExternalModule.php (hunk 1 above)
Comment (mmcev106) [reply FHIRServicesExternalModule.php]: I'm thinking about this a little differently.  The only case I know of where the value passed to `jsonSerialize()` or `toArray()` is expected to be `null` is on the `$response` in `send-record.php`.  In that light, I think it makes sense to put the `null` check in `send-record.php`, and leave the existing behavior everywhere else.  I believe [failing fast](https://en.wikipedia.org/wiki/Fail-fast) and crashing on `null` as soon as it is unexpected is a good thing.

Comment (j-frei) [reply FHIRServicesExternalModule.php]: Ah, I see. Yes, that approach makes sense as well.
I changed the code in the request where the check is performed in `send-record.php`.

Comment (mmcev106) [reply FHIRServicesExternalModule.php]: The latest changes look good to me.  Resolving!

**Comment by j-frei** [comment]
@mmcev106 Regarding the changes in [`send-record.php`](https://github.com/j-frei/fhir-services-module/blob/e23eafc17a1a277cbb27dc4818b28a09e488fa7e/send-record.php#L48):
At the moment, the pull request code does this:
```php
if(is_null($response)){
    $remote_response = "";
} else {
```
but you might prefer an encoding that is according to the JSON standard:
```php
if(is_null($response)){
    $remote_response = "null";
} else {
```
Which version do you prefer?

**Comment by mmcev106** [comment]
I already merged it!  I was considering suggesting not changing it (just leaving it as `null` as opposed to the string `"null"`), but returning the empty string is probably fine too.

**Comment by j-frei** [comment]
Alright. Thank you for your input and feedback!
